sdk/src/main/java/com/google/cloud/dataflow/sdk/coders/Proto2Coder.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (C) 2015 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.dataflow.sdk.coders;
+
+import com.google.protobuf.ExtensionRegistry;
+import com.google.protobuf.GeneratedMessage.GeneratedExtension;
+import com.google.protobuf.Message;
+import com.google.protobuf.Parser;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * An encoder using Google Protocol Buffers 2 binary format.
+ * <p>
+ * To learn more about Protocol Buffers, visit:
+ * https://developers.google.com/protocol-buffers
+ * <p>
+ * To use, specify the {@code Coder} type on a PCollection:
+ * <pre>
+ * {@code
+ * PCollection<MyProto.Message> records =
+ *     input.apply(...)
+ *          .setCoder(Proto2Coder.of(MyProto.Message.class));
+ * }
+ * </pre>
+ * <p>
+ * Custom message extensions are also supported, but the coder must be made
+ * aware of them explicitly:
+ * <pre>
+ * {@code
+ * PCollection<MyProto.Message> records =
+ *     input.apply(...)
+ *          .setCoder(Proto2Coder.of(MyProto.Message.class)
+ *          .withExtensionsFrom(MyProto.class));
+ * }
+ * </pre>
+ *
+ * @param <T> the type of elements handled by this coder, must extend {@code Message}
+ */
+public class Proto2Coder<T extends Message> extends CustomCoder<T> {
+
+  /**
+   * Produces a new Proto2Coder instance, for a given Protobuf message class.
+   */
+  public static <T extends Message> Proto2Coder<T> of(Class<T> protoMessageClass) {
+    return new Proto2Coder<>(protoMessageClass);
+  }
+
+  private final String protoMessageClassName;
+  private final List<String> extensionClassNames = new ArrayList<String>();
```

Comment:
Class is serializable, instead of storing the string name, just store the Class<?> and List<Class<?>>, saves you from having to look it up later with Class.forName